nntrainer/tensor/blas_interface.cpp

```diff
@@ -182,6 +186,13 @@ static void copy_float32_to_float16(const unsigned int N, const float *X,
     for (unsigned int i = 0; i < N; ++i)
       Y[i * incy] = X[i * incx];
   }
+#elif USE_AVX
+  if (incX == 1 && incY == 1) {
+    nntrainer::avx::vcvt_f32_f16(N, X, Y);
+  } else {
+    for (unsigned int i = 0; i < N; ++i)
+      Y[i * incy] = static_cast<_FP16>(X[i * incx]);
+  }
 #else
   for (unsigned int i = 0; i < N; ++i)
     Y[i * incy] = static_cast<_FP16>(X[i * incx]);
```

Comment:
This is not for this PR or this month, but for later versions (you are now having technical debts here):

maintain the same code in general cpp files across NEON/AVX/None.
centralize architectural dependencies to a corresponding header and its implementation only.

You may have a single header for all three cases and choose a corresponding implementation (.c/.cpp file) at build time, too.

Reply:
Try the better approach from the next implementation (and fix this when you are ready)

Reply:
You mean something like:

```C++
// new file : blas_raw.cpp
...
void foo( ... ) {
...
} else {
    for (unsigned int i = 0; i < N; ++i)
      Y[i * incy] = static_cast<_FP16>(X[i * incx]);
  }
...
}
...
```
and use it for all `blas_interface.cpp`, `blas_neon.cpp`, `blas_avx.cpp` ?
?

Reply:
Thank you for bringing this up! I will keep it in mind for future improvement.

Reply:
Or.. you may have a base class as a set of operators/functions/methods for general CPU operations compatible for ALL systems with virtual functions and a derived class as a set of operators/functions/methods for architecture-dependent (SIMD) operations overriding the virtual functions.

Then, if you don't have an implementation in the derived class (SIMD), the fallback method in the base class will be automatically chosen. If you have an implementation in the derived class and the class is chosen at run-time/build-time, it will be automatically chosen WITHOUT having #if/#endif or even general if statements. You can control that in the base class initializer or "get_instance" method of the singleton design.

Reply:
Such refactoring needs planning and design before implementation. Don't just start without enough discussion and I don't want this refacotring effort hinder the immenent release.

Anyway, for your reference: https://github.com/nnstreamer/nntrainer/discussions/2482

Reply:
The concept suggestion : https://github.com/myungjoo/nntrainer/tree/suggestion/refactoring/archdep
(this is not for actual PR)

Reply:
both approaches seem great. I'll also research how other frameworks manage this. let's keep the discussion going